text/2018-10-15-batch-client-messages.md

```diff
@@ -0,0 +1,69 @@
+
+# Detailed design
+
+## BatchCommands interface
+
+The new interface `BatchCommands` in `Tikv` service looks like this:
+```proto
+service Tikv {
+    // ... some old interfaces.
+    rpc BatchCommands(stream BatchCommandsRequest) returns (stream BatchCommandsResponse) {}
+}
+
+message BatchCommandsRequest {
+    repeated Request requests = 1;
+    repeated uint64 request_ids = 2;
+    
+    message Request {
+        oneof cmd {
+            GetRequest Get = 1;
+            ScanRequest Scan = 2;
+            // ... more other requests.
+        }
+    }
+}
+
+message BatchCommandsResponse {
+    repeated Response responses = 1;
+    repeated uint64 request_ids = 2;
+    // Indicates the TiKV's transport layer is in heavy load or not.
+    bool in_heavy_load = 3;
+
+    message Response {
+        oneof cmd {
+            GetResponse Get = 1;
+            ScanResponse Scan = 2;
+            // ... more other responses respectively.
+        }
+    }
+}
+```
+
+First of all, it's a dual-way streaming interface. It makes TiKV can be free to constuct responses for batched requests. For example, TiDB can send a `BatchCommandsRequest` with requests `[Req-1, Req-2, Req-3, Req-4, Req-5]`, however TiKV can give `BatchCommandsResponse`s back with responses `[Resp-1, Resp-5, Resp-3]` and `[Resp-2, Resp-4]`. It's unnecessary to wait for any responses which comes from one `BatchCommandsRequest`; the order just depends on processing speed of every requests.
+
+`request_ids` in both `BatchCommandsRequest` and `BatchCommandsResponse` is used to relate requests and responses in the streaming interface. When clients send a batch request to TiKV, they also need to store their `request_ids` somewhere, and then dispatch responses (extracted from batch responses) respectively to them.
+
+And then `oneof` is used to unify requests and responses, instead of a `message`. Their wired protocols are almost same, but the former's generated code is much more better.
+
+The last thing is `in_heavy_load` in `BatchCommandsResponse`, which is used for TiKV can tell clients the load of TiKV is heavy or not. So clients can adjust their strategy (e.g. add some backoff to avoid little batch) to be more effective for TiKV.
+
+## Implement in TiKV
+
+The implement in TiKV is very simle. It just extract `Request`s from `BatchCommandsRequest`, dispatch them to engine layer, and then collect the `Response`s into `BatchCommandsResponse` and send it to clients.
+
+## How TiKV knows its gRPC threads are busy
+
+TiKV gets CPU usage of gRPC threads by reading `/proc/<tikv-pid>/tasks/<grpc-tid>`, which is only avaliable on Linux. If gRPC CPU usage is greater than 80% in last 3 seconds, TiKV can tell clients it's in heavy load.
```

Comment:
you must notice that reading the file and parsing it are not cheap, use it carefully.

Reply:
Maybe you can get these values from prometheus metrics.

Reply:
How about utilizing gRPC metrics? They are more accurate and comprehensive.

Reply:
ping @hicqu